File: src/noiedigitalsystem/api/result/NdsResult.java
package noie.linmimeng.noiedigitalsystem.api.result;

import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * NDS Result - 結果模型（消滅 Exception）
 * 
 * <p>用於替代 Checked Exception，提供函數式錯誤處理。</p>
 * 
 * <p><b>憲法級規則：</b></p>
 * <ul>
 *   <li>API 層禁止 Checked Exception</li>
 *   <li>所有錯誤通過 NdsResult 表達</li>
 *   <li>支持函數式鏈式調用</li>
 * </ul>
 * 
 * @param <T> 成功時的數據類型
 * @since 2.0.0
 */
public interface NdsResult<T> {
    
    /**
     * 檢查是否成功
     * 
     * @return true 如果成功
     */
    boolean isSuccess();
    
    /**
     * 獲取數據
     * 
     * @return 數據（僅在成功時有效）
     * @throws IllegalStateException 如果結果不是成功
     */
    T data();
    
    /**
     * 獲取錯誤
     * 
     * @return 錯誤對象（僅在失敗時有效）
     * @throws IllegalStateException 如果結果不是失敗
     */
    NdsError error();
    
    /**
     * 映射成功值
     * 
     * @param mapper 映射函數
     * @param <R> 新類型
     * @return 新的結果
     */
    <R> NdsResult<R> map(Function<T, R> mapper);
    
    /**
     * 扁平映射成功值
     * 
     * @param mapper 映射函數（返回新的 Result）
     * @param <R> 新類型
     * @return 新的結果
     */
    <R> NdsResult<R> flatMap(Function<T, NdsResult<R>> mapper);
    
    /**
     * 處理成功情況
     * 
     * @param consumer 消費者
     * @return 當前結果
     */
    NdsResult<T> onSuccess(Consumer<T> consumer);
    
    /**
     * 處理失敗情況
     * 
     * @param consumer 消費者
     * @return 當前結果
     */
    NdsResult<T> onFailure(Consumer<NdsError> consumer);
    
    /**
     * 獲取 Optional 數據
     * 
     * @return Optional 數據
     */
    default Optional<T> toOptional() {
        return isSuccess() ? Optional.of(data()) : Optional.empty();
    }
    
    /**
     * 創建成功結果
     * 
     * @param data 數據
     * @param <T> 數據類型
     * @return 成功結果
     */
    static <T> NdsResult<T> success(T data) {
        return new NdsResultImpl<>(true, data, null);
    }
    
    /**
     * 創建失敗結果
     * 
     * @param error 錯誤對象
     * @param <T> 數據類型
     * @return 失敗結果
     */
    static <T> NdsResult<T> failure(NdsError error) {
        return new NdsResultImpl<>(false, null, error);
    }
    
    /**
     * 創建失敗結果（便捷方法）
     * 
     * @param code 錯誤代碼
     * @param message 錯誤消息
     * @param <T> 數據類型
     * @return 失敗結果
     */
    static <T> NdsResult<T> failure(String code, String message) {
        return failure(NdsError.of(code, message));
    }
}

File: csharp/src/Noie.Nds.Api.Abstractions/Result/NdsResult.cs
namespace Noie.Nds.Api.Result;

/// <summary>
/// [Index: NDS-CSHARP-NDSRESULT-000] NDS Result (exceptionless business outcomes).
/// 
/// <para>[Semantic] Functional success/failure envelope replacing checked exceptions.</para>
/// 
/// <para><b>Constitutional constraints:</b></para>
/// <list type="bullet">
///   <item>[Constraint] API layer must not rely on checked exceptions for business failures.</item>
///   <item>[Behavior] Business outcomes are expressed via NdsResult.</item>
///   <item>[Behavior] Supports functional chaining.</item>
/// </list>
/// </summary>
/// <typeparam name="T">Data type on success.</typeparam>
public readonly struct NdsResult<T>
{
    private readonly T? _data;
    private readonly NdsError? _error;
    
    /// <summary>
    /// Check whether this result is successful.
    /// </summary>
    public bool IsSuccess { get; }
    
    /// <summary>
    /// Check whether this result is a failure.
    /// </summary>
    public bool IsFailure => !IsSuccess;
    
    /// <summary>
    /// Get the success data.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the result is not successful.</exception>
    public T Data => IsSuccess 
        ? _data! 
        : throw new InvalidOperationException("Cannot access data on a failed result");
    
    /// <summary>
    /// Get the failure error.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the result is successful.</exception>
    public NdsError Error => !IsSuccess 
        ? _error! 
        : throw new InvalidOperationException("Cannot access error on a successful result");
    
    private NdsResult(bool isSuccess, T? data, NdsError? error)
    {
        IsSuccess = isSuccess;
        _data = data;
        _error = error;
    }
    
    /// <summary>
    /// Map the success value.
    /// </summary>
    /// <typeparam name="TResult">New result type.</typeparam>
    /// <param name="mapper">Mapping function.</param>
    /// <returns>New result.</returns>
    public NdsResult<TResult> Map<TResult>(Func<T, TResult> mapper)
    {
        return IsSuccess 
            ? NdsResult<TResult>.Success(mapper(_data!)) 
            : NdsResult<TResult>.Failure(_error!);
    }
    
    /// <summary>
    /// Flat-map the success value.
    /// </summary>
    /// <typeparam name="TResult">New result type.</typeparam>
    /// <param name="mapper">Mapping function (returns a new result).</param>
    /// <returns>New result.</returns>
    public NdsResult<TResult> FlatMap<TResult>(Func<T, NdsResult<TResult>> mapper)
    {
        return IsSuccess 
            ? mapper(_data!) 
            : NdsResult<TResult>.Failure(_error!);
    }
    
    /// <summary>
    /// Run an action when successful.
    /// </summary>
    /// <param name="action">Action.</param>
    /// <returns>The current result.</returns>
    public NdsResult<T> OnSuccess(Action<T> action)
    {
        if (IsSuccess) action(_data!);
        return this;
    }
    
    /// <summary>
    /// Run an action when failed.
    /// </summary>
    /// <param name="action">Action.</param>
    /// <returns>The current result.</returns>
    public NdsResult<T> OnFailure(Action<NdsError> action)
    {
        if (!IsSuccess) action(_error!);
        return this;
    }
    
    /// <summary>
    /// Get the data or a default value.
    /// </summary>
    public T? GetOrDefault(T? defaultValue = default)
    {
        return IsSuccess ? _data : defaultValue;
    }
    
    /// <summary>
    /// Match success or failure.
    /// </summary>
    public TResult Match<TResult>(Func<T, TResult> onSuccess, Func<NdsError, TResult> onFailure)
    {
        return IsSuccess ? onSuccess(_data!) : onFailure(_error!);
    }
    
    /// <summary>
    /// Create a successful result.
    /// </summary>
    public static NdsResult<T> Success(T data)
    {
        return new NdsResult<T>(true, data, null);
    }
    
    /// <summary>
    /// Create a failed result.
    /// </summary>
    public static NdsResult<T> Failure(NdsError error)
    {
        return new NdsResult<T>(false, default, error);
    }
    
    /// <summary>
    /// Create a failed result (convenience overload).
    /// </summary>
    public static NdsResult<T> Failure(string code, string message)
    {
        return Failure(NdsError.Of(code, message));
    }
    
    /// <summary>
    /// Implicit conversion from T to NdsResult&lt;T&gt; (success).
    /// </summary>
    public static implicit operator NdsResult<T>(T value) => Success(value);
    
    /// <summary>
    /// Implicit conversion from NdsError to NdsResult&lt;T&gt; (failure).
    /// </summary>
    public static implicit operator NdsResult<T>(NdsError error) => Failure(error);
}

/// <summary>
/// [Index: NDS-CSHARP-NDSRESULT-100] NdsResult helpers.
/// </summary>
public static class NdsResult
{
    /// <summary>
    /// Create a successful result.
    /// </summary>
    public static NdsResult<T> Success<T>(T data) => NdsResult<T>.Success(data);
    
    /// <summary>
    /// Create a failed result.
    /// </summary>
    public static NdsResult<T> Failure<T>(NdsError error) => NdsResult<T>.Failure(error);
    
    /// <summary>
    /// Create a failed result.
    /// </summary>
    public static NdsResult<T> Failure<T>(string code, string message) 
        => NdsResult<T>.Failure(code, message);
    
    /// <summary>
    /// Create an empty successful result.
    /// </summary>
    public static NdsResult<Unit> Ok() => NdsResult<Unit>.Success(Unit.Value);
}

/// <summary>
/// [Index: NDS-CSHARP-UNIT-000] Represents an empty value (void-like).
/// </summary>
public readonly struct Unit
{
    /// <summary>
    /// The single Unit value.
    /// </summary>
    public static readonly Unit Value = new();
    
    public override string ToString() => "()";
}
